### PR TITLE
#1144 - Part 1 : Basic concept and forward constraining wide

### PR DESCRIPTION
**Pull request by jfalcou**
The basic.hpp concept file defines : 
 - `plain_scalar` i.e float, ints etc
 -  `scalar_product_type ` which is a `kumi::product_type` of `plain_scalar` 
 - `scalar` which is `plain_scalar` or `scalar_product_type `

Added tests and made wide constrained on vectorizable.
Took the opportunity to make empty tuple non-vectorizable.

### CODE REVIEW

File: include/eve/concept/basic.hpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#include <type_traits>
+#include <cstdint>
+#include <cstddef>
+
+namespace eve
+{
+  template<typename T>
+  concept builtin_vectorizable  =     std::is_arithmetic_v<T>
+                                  && !std::is_same_v<T,bool>
+                                  &&  sizeof(T) <= 8;
```
Comment (DenisYaroshevskiy) [inline include/eve/concept/basic.hpp:20]: I don't like the name tbh. 

Given that is_integral is already a list, do we maybe want to just list everything?

https://github.com/llvm/llvm-project/blob/50de7f1e7787bf327d1d04c4daab97c1c180b58c/libcxx/include/__type_traits/is_integral.h#L22

**Review by DenisYaroshevskiy — COMMENTED** [review]
How about `plain_scalar` and `scalar`?

File: include/eve/concept/basic.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/concept/basic.hpp:20]: I fear we may miss some std::intx_t on some architecture if we use a list

Comment (DenisYaroshevskiy) [reply include/eve/concept/basic.hpp:20]: We should miss it. We don't support char for example. We have a very specific list.

Comment (jfalcou) [reply include/eve/concept/basic.hpp:20]: OK

File: include/eve/concept/scalar.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <eve/detail/kumi.hpp>
+#include <type_traits>
+#include <cstdint>
+#include <cstddef>
+
+namespace eve
+{
+  template<typename T>
+  concept plain_scalar  = std::is_same_v<T,float>         ||  std::is_same_v<T,double>
```
Comment (DenisYaroshevskiy) [inline include/eve/concept/scalar.hpp:18]: a) we should have one_of concept
b) we should use std::same_as concept

File: include/eve/concept/scalar.hpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        {
+          using flatten_t = kumi::result::flatten_all_t<T>;
+          return []<std::size_t... I>( std::index_sequence<I...> )
+          {
+            return (plain_scalar<kumi::element_t<I,flatten_t>> && ... && true);
+          }(std::make_index_sequence<kumi::size<flatten_t>::value>{});
+        }
+        else
+        {
+          return false;
```
Comment (DenisYaroshevskiy) [inline include/eve/concept/scalar.hpp:40]: Is no elelements a problem

File: include/eve/concept/scalar.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/concept/scalar.hpp:18]: a) we have detaill:is_on_of<T>( types<U...>)  available

File: include/eve/concept/scalar.hpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#include <eve/detail/kumi.hpp>
+#include <eve/detail/meta.hpp>
+#include <type_traits>
+#include <cstdint>
+#include <cstddef>
+
+namespace eve
+{
+  template<typename T, typename... Ts>
+  concept one_of = (std::same_as<T,Ts> || ... || false);
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/concept/scalar.hpp:19]: it helps if Ts... is empty

**Review by DenisYaroshevskiy — APPROVED** [review]

**Pull request merged**